// nodejs-tools/src/main.js
module.exports = (function (fs, path, clog) {
    // process log
    var console = {
        log: function () {
            var args = [].slice.apply(arguments);
            clog.apply(undefined, [ 'tools:' ].concat(args));
        }
    };

    var _pub_static = function () {
        var _pri = {};
        var _pub = {};

        var _init = function () {
            _pri.sourcePath = path.normalize(path.dirname(__dirname) + '/src');
            // console.log(_pri.sourcePath);
        };

        _pri['sourcePath'] = null;

        _pub['file'] = require('./file.helper.js');

        _pub['htmlToString'] = function (htmlStr) {
            var template = [];
            htmlStr = htmlStr || '';
            content = htmlStr.split(/\n|\r/);
            content.forEach(function (s) {
                var len;
                if (s.trim() === '') {
                    return;
                }

                s = s.replace(/\'/g, '\\\'').replace(/\"/g, '\\\"');
                template.push(s);
            });

            return template.join('');
        };

        _pub['htmlToJs'] = function (htmlStr) {
            var tmplPath = _pri.sourcePath + '/tmpl.seajs.js';
            var template = fs.readFileSync(tmplPath, 'utf8');

            template = template.split('/*push data*/');
            htmlStr = htmlStr || '';
            content = htmlStr.split(/\n|\r/);
            content.forEach(function (s) {
                var len;
                if (s.trim() === '') {
                    return;
                }

                s = s.replace(/\'/g, '\\\'').replace(/\"/g, '\\\"');
                len = template.length;
                template[len] = template[len - 1];
                template[len - 1] = '\n    templateArray.push(\'' + s + '\');';
            });

            return template.join('');
        };

        _pub['htmlToJsNs'] = function (htmlStr, ns) {
            var tmplPath = _pri.sourcePath + '/tmpl.namespace.js';
            var template = fs.readFileSync(tmplPath, 'utf8');

            template = template.split('/*push data*/');
            template[1] = template[1].replace('/*namespace*/', ns);
            htmlStr = htmlStr || '';
            content = htmlStr.split(/\n|\r/);
            content.forEach(function (s) {
                var len;
                if (s.trim() === '') {
                    return;
                }

                s = s.replace(/\'/g, '\\\'').replace(/\"/g, '\\\"');
                len = template.length;
                template[len] = template[len - 1];
                template[len - 1] = '\n    templateArray.push(\'' + s + '\');';
            });

            return template.join('');
        };

        _pub['watch'] = require('./watch.js');

        if (this === 'test') {
            _pub._pri = _pri;
            _pub._init = _init;
        } else {
            _init.apply(_pub, arguments);
        }

        return _pub;
    };

    return new _pub_static();
})(
    require('fs'),
    require('path'),
    console.log
);
